pkg/policy/bucket-policy.go

```diff
@@ -82,6 +84,33 @@ type User struct {
 	CanonicalUser set.StringSet `json:"CanonicalUser,omitempty"`
 }
 
+// UnmarshalJSON is a custom json unmarshaler for Principal field,
+// the reason is that Principal can take a json struct represented by
+// User string but it can also take a string.
+func (u *User) UnmarshalJSON(data []byte) error {
```

Comment:
As fields in `User` are `set.StringSet`, `set.StringSet` supports single string or string array as JSON value.   Could you tell me what exactly is not working?  As per commit log, I see this method behaves like in `set.StringSet`.

Reply:
Perhaps adding a test would be better here @vadmeste to demonstrated.

Reply:
Added tests @harshavardhana 

@balamurugana, if you can see test n°2, you can see `"Principal": "*"`, it wasn't supported before this PR. Only `"Principal": { "AWS": "xx", "CanonicalUser": "xx"}` is supported.

set.StringSet is not enough in this case, AFAIK, it can only support a string and an array of strings but not a struct (struct User in this case)

Reply:
@vadmeste understood now.

It would be better to have below approach which makes things clean.

1. Have a type `user` exactly same as `User`
2. Extend `user` as `User`
3. Have `UnmarshalJSON()` to `User`.

Reply:
> It would be better to have below approach which makes things clean.
> 
> Have a type user exactly same as User
> Extend user as User
> Have UnmarshalJSON() to User.

This is practically what is done in the current approach, anything wrong with it ?

Reply:
you are doing in reverse

Reply:
> you are doing in reverse

oh yes, understood, but if I do so, it means I need to declare "user" outside UnmarshalJSON(), however no code except UnmarshalJSON() needs it, that's why I did it this way.

Reply:
The fact is a type is buried inside `UnmarshalJSON()`.  Hope it won't confuse new comers.